### ISSUE TITLE
Sarama consumer dropping messages randomly

### ISSUE THREAD

**Issue by uohzxela** [Shopify/sarama#464]
We are using Heka as log transporter, which has Kafka input & output plugins that depend on Sarama library. KafkaOutput plugin uses Sarama producer library, and KafkaInput uses Sarama consumer library. I created an issue recently (https://github.com/mozilla-services/heka/issues/1559) in Heka repository and did further troubleshooting, turns out that the problem actually lies with the Sarama consumer channel. 

Since Heka is still using an outdated version of Sarama, I replaced their KafkaInput plugin with a version that's using the latest Sarama. What KafkaInput really does is to detect messages from the consumer channel in a select block, then send the payload using SplitBytes(): https://github.com/uohzxela/heka-kafka-input/blob/master/kafka_input.go#L242

To test if the messages are dropped by Sarama consumer or by Heka, I added print statement before calling Heka's SplitBytes() call (https://github.com/uohzxela/heka-kafka-input/blob/master/kafka_input.go#L246), and the results point to the fact that Sarama consumer does indeed drop messages randomly. 

Here's a self-explanatory video that shows the whole process (left terminal is print statements from the Sarama consumer range loop in KafkaInput plugin, right terminal is output from tailing the Kafka logs): https://www.dropbox.com/s/iekaewyy7lxdv0y/kafka_input_bug.mp4?dl=0

Surprisingly, KafkaOutput (which uses Sarama producer) works perfectly fine because in the video, we used Postman to send requests to an API server, which sends the request logs to a Kafka topic using KafkaOutput plugin.

**Comment by wvanbergen** [comment]
Some questions to dig into this some more:

- How many partitions does your topic have? I see you are [only consuming partition 0](https://github.com/uohzxela/heka-kafka-input/blob/master/kafka_input.go#L195), and you're losing about 50% of the messages.
- Can you print the `message.Offset`? The sequence of offsets in a single partition should not have any gaps in it, e.g. 10, 11, 12, 13, 14, .... If there's gaps, that means there's a problem on the consumer side. If not, the problem is on the producer side.
- Where can I found the code of the producer side? What settings does it use?

**Comment by eapache** [comment]
Per Willem's first point: this definitely looks to me like you are producing to multiple partitions, but only consuming from one of them.

**Comment by aranair** [comment]
:+1:

**Comment by uohzxela** [comment]
Thanks @wvanbergen @eapache  for the help! We do have 2 partitions, the problem lies with the fact that KafkaOutput plugin (Sarama producer) is producing to either one of the partitions randomly (the default setting for the partitioner is [Random](https://hekad.readthedocs.org/en/latest/config/outputs/kafka.html)), and we are only consuming from one partition. I changed the Kafka config to have only one partition and it works now. Closing this issue since the problem lies with our Kafka config, not with Sarama.
